### PR TITLE
Upgrade istio to 1.10.2 and label and restart system components to pick up new sidecars

### PR DESCRIPTION
**Pull request by mbenmclean**
This PR covers calling istioctl to install the new canary control plane and gateways and then labeling and restarting the system components in order to pick up the new sidecars.

# Checklist 

As the author of this PR, I have:

- [ ] Checked that I included or updated copyright and license notices in all files that I altered
- [ ] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: THIRD_PARTY_LICENSES.txt [hunk 1]
```diff
@@ -2622,4 +2622,4 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 
 ATTRIBUTION-HELPER-GENERATED:
-License file based on go.mod with md5 sum: 08c17102b13f4a4f8a42b40c836100fa
+License file based on go.mod with md5 sum: 951c74a8f212aaabf16793d2a08b1a50
```
Comment (michael-cico) [inline THIRD_PARTY_LICENSES.txt:2625]: Should this have changed?  I thought we added `istioctl` already, or did this change not get made when we merged the istioctl support?

File: application-operator/controllers/metricstrait/metricstrait_controller.go [hunk 2]
```diff
@@ -893,8 +893,10 @@ func useHTTPSForScrapeTarget(ctx context.Context, c client.Client, trait *vzapi.
 	if err := c.Get(ctx, client.ObjectKey{Namespace: "", Name: trait.Namespace}, namespace); err != nil {
 		return false, err
 	}
-	value, ok := namespace.Labels["istio-injection"]
-	if ok && value == "enabled" {
+	istioInjection, hasIstioInjection := namespace.Labels["istio-injection"]
+	_, hasIstioRevision := namespace.Labels["istio.io/rev"]
+	hasSidecar := (hasIstioInjection && istioInjection == "enabled") || hasIstioRevision
```
Comment (michael-cico) [inline application-operator/controllers/metricstrait/metricstrait_controller.go:898]: For next pass, maybe this should be named `requiresSidecar`.  It's a nit, but I think that's probably more what we're checking here right?

File: go.mod [hunk 3]
```diff
@@ -26,6 +26,7 @@ require (
 	go.uber.org/zap v1.16.0
 	golang.org/x/lint v0.0.0-20201208152925-83fdc39ff7b5
 	golang.org/x/tools v0.0.0-20201224043029-2b0845dc783e
+	gopkg.in/errgo.v2 v2.1.0
```
Comment (michael-cico) [inline go.mod:29]: Not sure what this is?  Did you already get approval to include it?

File: platform-operator/controllers/verrazzano/component/helm/helm_component.go [hunk 4]
```diff
@@ -183,7 +184,7 @@ func (h HelmComponent) PostInstall(log *zap.SugaredLogger, client clipkg.Client,
 // that is included in the operator image, while retaining any helm Value overrides that were applied during
 // install. Along with the override files in helm_config, we need to generate image overrides using the
 // BOM json file.  Each component also has the ability to add additional override parameters.
-func (h HelmComponent) Upgrade(log *zap.SugaredLogger, client clipkg.Client, ns string, dryRun bool) error {
+func (h HelmComponent) Upgrade(log *zap.SugaredLogger, _ *installv1alpha1.Verrazzano, client clipkg.Client, ns string, dryRun bool) error {
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/helm/helm_component.go:187]: The arg is unused... is this for the next round or something?

Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/helm/helm_component.go:187]: I'm reworking this interface so we have a context object instead of an ever-expanding list of arguments, I can fix that on my next pass.

Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/helm/helm_component.go:187]: never mind, this is just implementing the contract change...I'll be changing it up later after you merge.

**Comment by michael-cico** [comment]
Make sure you squash the commits, either yourself or on merge.  Probably edit down the comments too.

File: platform-operator/controllers/verrazzano/component/istio/istio_component.go [hunk 5]
```diff
@@ -75,3 +136,111 @@ func (i IstioComponent) PreInstall(log *zap.SugaredLogger, client clipkg.Client,
...
+		// for Update call
+		deployment := deploy
+		if contains(i.InjectedSystemNamespaces, deployment.Namespace) {
+			if deployment.Spec.Paused {
+				return errors.Newf("Deployment %v can't be restarted because it is paused", deployment.Name)
+			}
+			if deployment.Spec.Template.ObjectMeta.Annotations == nil {
+				deployment.Spec.Template.ObjectMeta.Annotations = make(map[string]string)
+			}
+			deployment.Spec.Template.ObjectMeta.Annotations["verrazzano.io/restartedAt"] = time.Now().Format(time.RFC3339)
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/istio/istio_component.go:185]: It's good to see we can use our own prefix here -- I was a little reticent to use the `kubectl` prefix.

File: platform-operator/controllers/verrazzano/component/istio/istio_component.go [hunk 6]
```diff
@@ -75,3 +136,111 @@ func (i IstioComponent) PreInstall(log *zap.SugaredLogger, client clipkg.Client,
...
+			return err
+		}
+		log.Infof("Relabeled namespace %v for istio upgrade", platformNS.Name)
+	}
+	return nil
+}
+
+// restartSystemNamespaceResources restarts all the deployments, StatefulSets, and DaemonSets
+// in all of the Istio injected system namespaces
+func (i IstioComponent) restartSystemNamespaceResources(log *zap.SugaredLogger, client clipkg.Client) error {
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/istio/istio_component.go:168]: Not for this pass, but is there any indicator of what is or isn't in the mesh, or needs the sidecars?  It might be good to minimize the restarts here, especially in the system namespace where we know what needs to be restarted.

Since upgrade here is currently blocking and sequential for the time being, it would be good to minimize the restart times.

Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/istio/istio_component.go:168]: Actually, I think once you update the labels the restarts are asynchronous.  We'll have to do some steady-state monitoring and reporting on the restarted components at some point.

File: platform-operator/thirdparty/charts/weblogic-operator/values.yaml [hunk 7]
```diff
@@ -63,7 +63,7 @@ domainNamespaces:
 enableClusterRoleBinding: false
 
 # image specifies the container image containing the operator.
-image: "ghcr.io/oracle/weblogic-kubernetes-operator:3.3.0"
+image: "phx.ocir.io/weblogick8s/weblogic-kubernetes-operator:istio-11"
```
Comment (michael-cico) [inline platform-operator/thirdparty/charts/weblogic-operator/values.yaml:66]: This needs to be removed.  We shouldn't be pushing private registry paths into the public github repos.

Comment (michael-cico) [reply platform-operator/thirdparty/charts/weblogic-operator/values.yaml:66]: Actually it's the 3rd party chart, just revert it.

File: platform-operator/verrazzano-bom.json [hunk 8]
```diff
@@ -355,12 +355,13 @@
       "name": "weblogic-operator",
       "subcomponents": [
         {
-          "repository": "oracle",
+          "registry": "phx.ocir.io",
```
Comment (michael-cico) [inline platform-operator/verrazzano-bom.json:358]: same here, but I guess you needed it for the pipeline testing.

File: THIRD_PARTY_LICENSES.txt (hunk 1 above)
Comment (mbenmclean) [reply THIRD_PARTY_LICENSES.txt:2625]: I believe this caused a failure in the pipeline because this was wrong so I switched it. Is this just a unique string associated with a version of go.mod? I'm not sure what this does.

File: application-operator/controllers/metricstrait/metricstrait_controller.go (hunk 2 above)
Comment (mbenmclean) [reply application-operator/controllers/metricstrait/metricstrait_controller.go:898]: Ok I can change

File: go.mod (hunk 3 above)
Comment (mbenmclean) [reply go.mod:29]: I think it's to create an error. Maybe I'm not using the right package. I can check

File: platform-operator/controllers/verrazzano/component/istio/istio_component.go (hunk 5 above)
Comment (mbenmclean) [reply platform-operator/controllers/verrazzano/component/istio/istio_component.go:185]: Yes I don't think it matters about the prefix. I'm not even sure that the restarted part matters I think just any change to the annotation restarts the deployment. I'm not sure though I haven't tested with a generic annotation.

File: platform-operator/controllers/verrazzano/component/istio/istio_component.go (hunk 6 above)
Comment (mbenmclean) [reply platform-operator/controllers/verrazzano/component/istio/istio_component.go:168]: Yes I agree

File: go.mod (hunk 3 above)
Comment (michael-cico) [reply go.mod:29]: You can also use `fmt.Errorf` to create an error.

File: platform-operator/controllers/verrazzano/component/istio/istio_component_test.go [hunk 9]
```diff
@@ -4,54 +4,183 @@
...
+	"context"
+	"github.com/golang/mock/gomock"
 	"github.com/stretchr/testify/assert"
+	installv1alpha1 "github.com/verrazzano/verrazzano/platform-operator/apis/verrazzano/v1alpha1"
+	"github.com/verrazzano/verrazzano/platform-operator/constants"
 	"github.com/verrazzano/verrazzano/platform-operator/internal/config"
 	"github.com/verrazzano/verrazzano/platform-operator/internal/istio"
+	"github.com/verrazzano/verrazzano/platform-operator/mocks"
 	"go.uber.org/zap"
+	"gopkg.in/errgo.v2/fmt/errors"
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/istio/istio_component_test.go:16]: You don't need this, you can use `fmt.Errorf` to create errors.

File: platform-operator/thirdparty/charts/weblogic-operator/values.yaml (hunk 7 above)
Comment (mbenmclean) [reply platform-operator/thirdparty/charts/weblogic-operator/values.yaml:66]: Ack

File: platform-operator/verrazzano-bom.json (hunk 8 above)
Comment (mbenmclean) [reply platform-operator/verrazzano-bom.json:358]: Ack

File: platform-operator/controllers/verrazzano/component/istio/istio_component_test.go (hunk 9 above)
Comment (mbenmclean) [reply platform-operator/controllers/verrazzano/component/istio/istio_component_test.go:16]: Fixed

File: platform-operator/controllers/verrazzano/component/istio/istio_component.go [hunk 10]
```diff
@@ -30,25 +72,76 @@ func (i IstioComponent) IsOperatorInstallSupported() bool {
...
+			return err
+		}
+
+		if _, err = tmpFile.Write([]byte(istioOperatorYaml)); err != nil {
+			log.Errorf("Failed to write to temporary file: %v", err)
+			return err
+		}
+
+		// Close the file
+		if err := tmpFile.Close(); err != nil {
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/istio/istio_component.go:105]: need a defer to delete the tmp file

File: platform-operator/controllers/verrazzano/component/istio/istio_component.go [hunk 11]
```diff
@@ -30,25 +72,76 @@ func (i IstioComponent) IsOperatorInstallSupported() bool {
...
+	_, _, err = upgradeFunc(log, i.ValuesFile, tmpFile.Name())
+	if err != nil {
+		return err
+	}
+
+	err = labelAndResartFn(log, err, i, client)
+	if err != nil {
+		return err
+	}
+
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/istio/istio_component.go:122]: At some point we need to have the new CP be the default.  When are we going to do this?
 istioctl tag set default --revision 1-10-2. (or whatever the version is)

File: platform-operator/controllers/verrazzano/component/istio/istio_component.go (hunk 10 above)
Comment (mbenmclean) [reply platform-operator/controllers/verrazzano/component/istio/istio_component.go:105]: Right above this there is this code
defer os.Remove(tmpFile.Name())
Doesn't that do this?

File: platform-operator/controllers/verrazzano/upgrade.go [hunk 12]
```diff
@@ -18,7 +18,7 @@ import (
 )
 
 // The max upgrade failures for a given upgrade attempt is 2
-const failedUpgradeLimit = 2
+const failedUpgradeLimit = 200000
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/upgrade.go:21]: Do we really want this set this high?  If it's effectively unbounded then we should probably make a failed state non-terminal.  That is, we just keep trying to upgrade.

what made you set it this high?

Comment (mbenmclean) [reply platform-operator/controllers/verrazzano/upgrade.go:21]: I'm trying to debug OLCNE this will get changed back

File: platform-operator/controllers/verrazzano/controller.go [hunk 13]
```diff
@@ -122,7 +122,7 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	// Process CR based on state
 	switch vz.Status.State {
 	case installv1alpha1.Failed:
-		return r.FailedState(vz, log)
+		return r.ReadyState(vz, log)
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/controller.go:125]: Is this to make a failed state non-terminal (re-tryable)?

File: pkg/bom/bom_test.go [hunk 14]
```diff
@@ -226,29 +225,3 @@ func validateImages(assert *assert.Assertions, bom *Bom, checkImageVal bool) {
 		}
 	}
 }
-
```
Comment (michael-cico) [inline pkg/bom/bom_test.go:229]: why is this removed?

File: pkg/bom/bom.go [hunk 15]
```diff
@@ -62,9 +62,6 @@ type BomSubComponent struct {
 	// where ghci.io is the registry and verrazzano is the repository name.
 	Repository string `json:"repository"`
 
-	// Override the registry within a subcomponent
```
Comment (michael-cico) [inline pkg/bom/bom.go:65]: why is this removed?

Comment (michael-cico) [reply pkg/bom/bom.go:65]: I think we want this feature in the BOM package, it will allow us to override images for testing like you did.

File: platform-operator/controllers/verrazzano/upgrade.go (hunk 12 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/upgrade.go:21]: ah, OK sorry

File: pkg/bom/bom.go (hunk 15 above)
Comment (mbenmclean) [reply pkg/bom/bom.go:65]: Oh I thought you said remove that for now. I can add it back in

Comment (michael-cico) [reply pkg/bom/bom.go]: No, I meant just remove the BOM file overrides to pull it from the OCIR location... the actual functionality to allow those kind of registry overrides I wanted to keep as a feature of the BOM pkg.

File: platform-operator/controllers/verrazzano/controller.go (hunk 13 above)
Comment (mbenmclean) [reply platform-operator/controllers/verrazzano/controller.go:125]: Yes

File: platform-operator/controllers/verrazzano/component/istio/istio_component.go [hunk 16]
```diff
@@ -75,3 +165,142 @@ func (i IstioComponent) PreInstall(log *zap.SugaredLogger, client clipkg.Client,
...
+	}
+	for index := range statefulSetList.Items {
+		statefulSet := &statefulSetList.Items[index]
+
+		// Check if StatefulSet is in an Istio injected system namespace
+		if contains(i.InjectedSystemNamespaces, statefulSet.Namespace) {
+			if statefulSet.Spec.Template.ObjectMeta.Annotations == nil {
+				statefulSet.Spec.Template.ObjectMeta.Annotations = make(map[string]string)
+			}
+			statefulSet.Spec.Template.ObjectMeta.Annotations["verrazzano.io/restartedAt"] = time.Now().Format(time.RFC3339)
```
Comment (mbenmclean) [inline platform-operator/controllers/verrazzano/component/istio/istio_component.go:212]: change to constant

**Review by pfmackin — APPROVED** [review]
